indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java

```diff
@@ -0,0 +1,636 @@
+        // checkState throws IllegalStateException
+        checkState(!Strings.isNullOrEmpty(name), "Rule name is missing");
+        checkState(!Strings.isNullOrEmpty(filterTypeConfig),
+            "Rule filter type is missing: " + name);
+        checkState(!Strings.isNullOrEmpty(filterPatternConfig),
+            "Rule filter pattern is missing: " + name);
+        checkState(!Strings.isNullOrEmpty(actionConfig), "Rule action is missing: " + name);
+
+        // valueOf throws IllegalArgumentException
+        Action action = Action.valueOf(actionConfig.toUpperCase());
+        FilterType filterType = FilterType.valueOf(filterTypeConfig.toUpperCase());
+
+        switch (filterType) {
+          case REGEX:
+            checkState(!Strings.isNullOrEmpty(itemTypeConfig), "Item type is missing: " + name);
+            ItemType itemType = ItemType.valueOf(itemTypeConfig.toUpperCase());
+            if (!allowedItemTypes.contains(itemType)) {
+              throw new IllegalArgumentException(itemTypeConfig);
+            }
+            return new Rule(FilterType.REGEX,
+                name, Optional.of(itemType), action, new RegexPredicate(filterPatternConfig));
+          case FILE_PREFIX:
+            checkState(itemTypeConfig == null, "Item type should not be set for prefix rules");
+            return new Rule(FilterType.FILE_PREFIX, name, Optional.empty(),
+                action, new FilePrefixPredicate(action, filterPatternConfig));
+          case URL_PREFIX:
+            checkState(itemTypeConfig == null, "Item type should not be set for prefix rules");
+            return new Rule(FilterType.URL_PREFIX, name, Optional.empty(),
+                action, new UrlPrefixPredicate(action, filterPatternConfig));
+          default:
+            throw new IllegalArgumentException(filterTypeConfig);
+        }
+      }
+    }
+  }
+
+  private static class RegexPredicate implements Predicate<String> {
+    private final Pattern pattern;
+    private final String regex;
+
+    private RegexPredicate(String regex) throws PatternSyntaxException {
+      this.pattern = Pattern.compile(regex, Pattern.CASE_INSENSITIVE);
+      this.regex = regex;
+    }
+
+    @Override
+    public boolean apply(@Nullable String input) {
+      if (input == null) {
+        return false;
+      }
+      Matcher matcher = pattern.matcher(input);
+      return matcher.find();
+    }
+
+    @Override
+    public String toString() {
+      return regex;
+    }
+  }
+
+  private static class FilePrefixPredicate implements Predicate<String> {
+    private final Action action;
+    private final String configuredPrefix;
+    private final Path prefixPath;
+    private final Path prefixRoot;
+    private final List<Path> prefixComponents;
+
+    private FilePrefixPredicate(Action action, String prefix) {
+      this.action = action;
+      configuredPrefix = prefix;
+      prefixPath = Paths.get(prefix.toLowerCase());
```

Comment:
If we don't use toLowerCase, this predicate is no longer uniformly case-insensitive like the other predicates; it will depend on the platform where the filter's running. Files.isSameFile is the same as Path.equals when the paths are equal, and if not, according to the Javadoc, may need "to open or access both files", which seemed like more than this class wanted to do.